py/desiutil/timer.py

```diff
@@ -0,0 +1,125 @@
+"""
+timer: standardizing reporting of algorithm and I/O timing
+
+The `timer.Timer` class is intended for reporting timing of events that
+take seconds to minutes; it is not intended for detailed profiling.  Example::
+
+    from desiutil.timer import Timer
+    t = Timer()
+    t.start('blat')
+
+    #- Use context manager for timing simple steps, e.g. one-liners
+    with t.time('blat.input'):
+        stuff = io.read(filename)
+
+    #- Or use full start/stop if you prefer
+    t.start('blat.algorithm')
+    results = calculate(stuff)
+    t.stop('blat.algorithm')
+
+    with t.time('blat.output'):
+        io.write(results)
+
+    #- Stop outer timer; Note that named timers can be nested or interleaved
+    t.stop('blat')
+
+    #- Print a json report of the timing
+    print(t.report())
+"""
+
+import traceback
+import time
+import datetime
+import json
+import copy
+import os.path
+from contextlib import contextmanager
+
+class Timer(object):
+    """
+    A basic timer class for standardizing reporting of algorithm and I/O timing
+    
+    TIMER:<START|STOP>:<filename>:<lineno>:<funcname>: <message>
+    """
+    
+    def __init__(self):
+        self.timers = dict()
+    
+    def _prefix(self, step):
+        """
+        Return standard prefix string for timer reporting
+        
+        Args:
+            step (str): timing step, e.g. "START" or "STOP"
+        """
+        stack = traceback.extract_stack()
+        #- Walk backwards in stack to find first caller not from this file
+        #- and not contextlib.py
+        #- (exact index depends on whether context manager was used or not)
+        thisfile = os.path.normpath(__file__)
+        for caller in stack[-1::-1]:
+            if os.path.normpath(caller.filename) != thisfile and \
+               os.path.basename(caller.filename) != 'contextlib.py':
+                break
+        
+        filename = os.path.basename(caller.filename)
+        return f"TIMER-{step}:{filename}:{caller.lineno}:{caller.name}:"
+    
+    def start(self, name):
+        """Start a timer `name` (str)"""
+        # timestamp = datetime.datetime.now().replace(microsecond=0).isoformat()        
+        timestamp = datetime.datetime.now().isoformat()        
+        if name in self.timers:
+            print(self._prefix('WARNING'), f'Restarting {name} at {timestamp}')
```

Comment:
I'm very concerned that you are using bare `print()` statements here.  If the intention is to *only* use this in a laptop command-line or Jupyter notebook environment, maybe that's fine.  But it is not clear from the PR comments or the documentation in the module itself what the intended use environment is.

You do say that you're deliberately not using `desiutil.log`, but what if you need the timing in something that is sending output to a log file?  Yes `desiutil.log` writes to stdout, but that is also often overridden.  Furthermore `desiutil.log` may buffer output differently even when writing to stdout, resulting in log messages and timing messages being written out of order.

Reply:
One more thought:  if Jupyter notebooks *are* an intended environment for this, document why exactly one would want to use this as opposed to Jupyter built-ins like `%time` and `%%time`.

Reply:
Do you have suggestions for how to get this to play nicely with a logger while still having it print the filenames and line numbers for the code that called the timer, instead of for the line numbers within `timer.py` that call the logger?

e.g. if I have a script `blat.py`:
```python
import time
from desiutil.timer import Timer
t = Timer()
t.start('blat')
time.sleep(0.1)
t.stop('blat')
```
The current timer code produces:
```
TIMER-START:blat.py:4:<module>: Starting blat at 2020-09-09T14:25:54.262900
TIMER-STOP:blat.py:6:<module>: Stopping blat at 2020-09-09T14:25:54.364106; duration 0.10 seconds
```
But if I replace the bare print statements with log.info calls, I get:
```
INFO:timer.py:84:start: Starting blat at 2020-09-09T14:25:04.315599
INFO:timer.py:106:stop: Stopping blat at 2020-09-09T14:25:04.420473; duration 0.10 seconds
```
i.e. it preserves the fact that timers are starting and stopping, but not what file/lineno started/stopped them, which is what I also want in the messages, but I'm not sure how to do that with a logger.

Reply:
I don't have any off-the-cuff suggestions.  Answering that would require some in-depth research, not something you can do in a few minutes.

Real-world testing of the existing code might prove to be more important here.  For example, if this turns out this is fundamentally broken in a multiprocessing/mpi/openmp/... environment, then maybe it's OK to keep the print statements, since it would only ever work in a small command-line script.